packages/runtime-core/src/components/Teleport.ts

```diff
@@ -230,28 +230,27 @@ export const TeleportImpl = {
     parentSuspense: SuspenseBoundary | null,
     optimized: boolean,
     { um: unmount, o: { remove: hostRemove } }: RendererInternals,
-    doRemove: Boolean
+    doRemove: boolean
   ) {
     const { shapeFlag, children, anchor, targetAnchor, target, props } = vnode
 
     if (target) {
       hostRemove(targetAnchor!)
     }
 
-    // an unmounted teleport should always remove its children if not disabled
-    if (doRemove || !isTeleportDisabled(props)) {
-      hostRemove(anchor!)
-      if (shapeFlag & ShapeFlags.ARRAY_CHILDREN) {
-        for (let i = 0; i < (children as VNode[]).length; i++) {
-          const child = (children as VNode[])[i]
-          unmount(
-            child,
-            parentComponent,
-            parentSuspense,
-            true,
-            !!child.dynamicChildren
-          )
-        }
+    // an unmounted teleport should always unmount its children whether it's disabled or not
```

Comment:
this will be a breaking change.

Reply:
The original comment was added in #2870. And it refered to `remove` but not `unmount`. 
When it comes to `remove`, there is indeed no need to remove its children when it's disabled and `doRmove` is false.
But `unmount` is not equal to `remove`. We can unmount a child without remove it, by making `doRemove` false.

While unmounting a component, we should also unmount its children. If not, its children would never be unmounted, and will cause the bug this pr fixes.